Having rotated accelerations use quaternions instead of gyroscope

wlsanderson opened the pull request:
Because we have good datasets now, we can have the rotated accelerations be calculated with quaternions directly and not from gyro. We will have gyro as a backup, and if any of the uncertainties of the quaternions go above some threshold (in constants file), the data processor will calculate with gyroscope as a fallback. This change should improve the speed of this function by a lot, as it can vectorize a large portion of it.

JacksonElia commented:
Do you think that the quaternions will be more accurate than using the gyroscope? I'm not too well versed in this but don't the quaternions have the potential to accumulate error? I guess because its just accel though it doesn't really matter.

harshil21 commented on airbrakes/data_handling/data_processor.py:
```diff
@@ -202,52 +202,111 @@ def _calculate_current_altitudes(self) -> npt.NDArray[np.float64]:
-        # Iterates through the data points and time differences between the data points
-        for idx, (data_packet, dt) in enumerate(
-            zip(self._data_packets, self._time_differences, strict=True)
+        # checks for missing data
+        if any(
+            packet.estCompensatedAccelX is None
+            or packet.estCompensatedAccelY is None
+            or packet.estCompensatedAccelZ is None
+            for packet in self._data_packets
```
```suggestion
```
fair to say if any of these are missing we are cooked. Better to just not include it it then.

wlsanderson replied:
Yeah I was kinda debating this idea with what to check when running this function. It's not like we have ever encountered a situation where a packet doesn't have acceleration or gyroscope, so idk if it is even worth checking

harshil21 replied:
Yeah we have enough launches with this IMU to know that's not going to happen

harshil21 commented on airbrakes/data_handling/data_processor.py:
```diff
@@ -202,52 +202,111 @@ def _calculate_current_altitudes(self) -> npt.NDArray[np.float64]:
-            # during motor burn is positive, and acceleration due to drag force during coast phase
-            # is negative.
-            rotated_accelerations[idx] = -rotated_accel[2]
+            return rotated_accelerations
```
```suggestion
```
see above. github doesn't allow me to give suggestions on deleted lines so

harshil21 commented on airbrakes/data_handling/data_processor.py:
```diff
@@ -202,52 +202,111 @@ def _calculate_current_altitudes(self) -> npt.NDArray[np.float64]:
+
+        if (
+            np.any(quaternion_uncertainties >= QUATERNION_UNCERTAINTY_THRESHOLD)
+            or self._data_packets[-1].estOrientQuaternionW is None
```
why is only the W quaternion checked for None? To account for missing quaternions in interest launch data? Then imo we should modify that dataset also, to add a random value, maybe exactly 0.0. We should also add `invalid_fields` for that dataset.

wlsanderson replied:
yeah I'll go back and flesh it out a bit better. You think I should just check for missing quaternions and missing quaternion uncertainties only? Don't really know if checking for missing accel/gyro is worth it

harshil21 replied:
I don't think we should check any of that, they are always guaranteed to be there. Let's just modify the older datasets

harshil21 commented on airbrakes/data_handling/data_processor.py:
```diff
@@ -202,52 +202,111 @@ def _calculate_current_altitudes(self) -> npt.NDArray[np.float64]:
+            np.any(quaternion_uncertainties >= QUATERNION_UNCERTAINTY_THRESHOLD)
+            or self._data_packets[-1].estOrientQuaternionW is None
+        ):
+            # quaternions are too uncertain to use, so we calculate them from gyro
```
calculating from gyro data should be in a different method

wlsanderson replied:
Gotcha, should this logic of determining which calculation to use be in the data_processor update method? or should the update method call quaternion rotation method, and that will call the gyro rotation method if uncertainty is too high?

harshil21 replied:
the latter. the update() method is a higher level method and doesn't need to be concerned with the specifics

harshil21 commented on airbrakes/data_handling/data_processor.py:
```diff
@@ -202,52 +202,111 @@ def _calculate_current_altitudes(self) -> npt.NDArray[np.float64]:
+            ]
+        )
+        # rotate each compensated acceleration by the most recent quaternion orientation. This
+        # introduces some slight error, because the orientation for the last packet is not going
+        # to be the orientation for each packet, but it should be a negligible difference.
```
after #99 the error would be even lower, but I would be interested in what the overall difference in predicted apogee is, when running the sim with `-f`.

harshil21 commented on tests/test_data_processor.py:
```diff
@@ -349,6 +401,10 @@ def test_first_update_no_data_packets(self, data_processor):
+                        estAttitudeUncertQuaternionW=0.01,
+                        estAttitudeUncertQuaternionX=0.01,
+                        estAttitudeUncertQuaternionY=0.01,
+                        estAttitudeUncertQuaternionZ=0.01,
```
lol we really need a helper method to generate data packets like we have in test_logger, this must've been atrocious to edit

harshil21 commented on tests/test_data_processor.py:
```diff
@@ -499,11 +555,58 @@ def test_max_altitude(self, data_processor):
+        rotations = d._rotated_accelerations
+        assert len(rotations) == n_packets
+
+        assert rotations[-1] == pytest.approx(expected_value)
```
Interesting that the expected_value is the same for both methods. I would've expected a small difference at least.

harshil21 commented on tests/test_data_processor.py:
```diff
@@ -499,11 +555,58 @@ def test_max_altitude(self, data_processor):
         rotations = d._rotated_accelerations
         assert len(rotations) == n_packets
 
         assert rotations[-1] == pytest.approx(expected_value)
```
```suggestion
        assert rotations[-1] == pytest.approx(expected_value)
        assert rotations[0] == rotations[-1]
```
this checks that all the rotations set were the same for that list for the quaternion only case.

harshil21 reviewed (changes requested):
I think we discussed this briefly, but wouldn't it be safer to just check for `invalid_fields` rather than uncertainties?

harshil21 commented on airbrakes/data_handling/data_processor.py:
```diff
@@ -202,52 +202,111 @@ def _calculate_current_altitudes(self) -> npt.NDArray[np.float64]:
-
-        return rotated_accelerations
+        # we only want to return the negative vertical component of each acceleration
+        return np.array([-accel[2] for accel in rotated_accels])
```
```suggestion
        return -rotated_accels[:, -1]
```
This is **95.1%** faster since it's vectorized!!

```python
import timeit

setup = """
import numpy as np

n = np.array([[1.213, 1.213 + i, 1.213 +2*i] for i in range(100)])
"""

test_1 = """
b = -n[:, -1]
"""

test_2 = """
b = np.array([-j[-1] for j in n])
"""

time_1 = timeit.timeit(test_1, setup, number=1000000)
time_2 = timeit.timeit(test_2, setup, number=1000000)

print(time_1)
print(time_2)
```
```bash
└─[$] <git:(adjust-queue-fetching*)> python scripts/perf.py
0.6513239109990536
24.606471167000564
```

wlsanderson replied:
man I was sure there was a way to vectorize this but couldn't figure it out lol

harshil21 replied:
https://numpy.org/devdocs/user/basics.rec.html#structured-arrays

oh my god look what I found

wlsanderson replied:
if we did this, would msgspec not be needed?

harshil21 replied:
it's a bit early to say that, but I'm looking at alternative ways of representing data, like pandas Dataframes (which are also labeled). It's just that it hurts the readability of having data packets to represent data v/s Dataframes, but the latter is way more suitable for us since we keep working on the columns (e.g. data processor and apogee predictor) rather than the row (only used in logging)

harshil21 replied:
Resolved in #112

wlsanderson commented:
> I think we discussed this briefly, but wouldn't it be safer to just check for `invalid_fields` rather than uncertainties?

No, the issue with invalid fields is that it is not flexible enough. From the previous datasets, it seemed that quaternions became invalid when an uncertainty component was around 0.1, but usually stayed invalid for multiple seconds, even when the uncertainty dropped back significantly below 0.1. Being able to tweak this value ourselves by just using the uncertainty is much more versatile. 0.1 is also incredibly sensitive, and I think increasing this to 0.5 is still well within a safe margin. We have only seen issues with quaternion uncertainty when it had that bug and hit a billion.

Even in freefall when some of the components of quaternion uncertainty reached over 2.0, the data is still good enough for us to get accurate (+/- 20 m/s) velocity tracking once it hits the ground.

wlsanderson commented:
> Do you think that the quaternions will be more accurate than using the gyroscope? I'm not too well versed in this but don't the quaternions have the potential to accumulate error? I guess because its just accel though it doesn't really matter.

I'm not exactly sure how the IMU does this, but the quaternions are absolute measurements, not relative. Basically, the quaternions always give the rotation with respect to some fixed axis, regardless of what the initial orientation of the IMU was when it was powered on. Because of this, I'd imagine quaternions are less susceptible to gyro drift. Gyroscope is relative, and so we just have to use dead reckoning to estimate our orientation.

wlsanderson commented:
Still a work in progress, but it is possible that our gyroscope calculations were wrong the entire time. I don't really know how. I added quaternion uncertainty to purple nurple, and added the missing quaternions to interest launch. When checking interest launch, it turns out that during the beginning of coast phase with gyroscope calculations that we are currently doing on the main branch, we are at a small pitch angle of around 5-10 degrees. The original quaternions (the ones that actually logged in the original flight) are at a pitch angle of around 20 - 30 degrees at the same point in flight.

I'm not actually sure which one is correct. Its possible that the gyroscope data does not line up with the orientation of the rocket, or it is possible that in the interest launch data when quaternion data became "valid" again, it had a large amount of error. I will do some checking with genesis flights to see how those compare with pure quaternion  vs pure gyro, but it might be different enough between full-scale and sub-scale that genesis won't be a good comparison. If this is true though, this might be the reason why #76 apogee prediction convergence time doesn't work for only full scale, but not sub scale. Thoughts? @JacksonElia @harshil21

wlsanderson commented:
After thinking about it, i feel like I might know why this is. I'm not sure which one, but I think either integrating gyro, or IMU quaternions, actually give the orientation of the rocket (which way it is facing), and one of them gives the direction it is travelling (direction of the acceleration vector). The angle between these two might be the angle of attack of the rocket

harshil21 commented:
Interesting observations. This is why I wanted some kind of a comparison between the two approaches. Do you think that integrating gyro data (which is prone to drift) is causing a large error between the two? What does the plot of the pitch from T+0 to some time in coast phase look like for gyro data only v/s quaternion data only? (I guess only for genesis since we actually logged that).

I would bias towards quaternions being more accurate since that's what they actually represent the absolute orientation.

wlsanderson commented:
I'm going to test all that stuff today.

> I would bias towards quaternions being more accurate since that's what they actually represent the absolute orientation.

That's what I originally thought too, but I dm'd Sailor last night and he said that the reason it seems absolute is most likely because it uses something like the gravity vector to initialize the quaternion, and then iterate on that with gyroscope. So it's "absolute" for the first packet when the IMU boots up, and then uses that as a reference axis for relative measurements afterwards.

I'm going to test more though to see what is most reasonable. I definitely want to consider getting a standalone magnetometer though, for the IMU to use as an external heading system.

harshil21 commented:
If you/we have time, we can go over the mscl library code, then we'd actually know what they are doing and e.g. when quaternions become invalid instead of just guessing

wlsanderson commented:
sure, i'm available all today in the lab to work on it

wlsanderson commented:
Closing this PR because after reviewing interest launch, and both genesis flight data, I believe that determining orientation from gyroscope is still the most robust and consistent method. The branch will remain available because we might return to it in the future, if a magnetometer proves to fix this issue.

I graphed the pitch versus time for interest launch, taking note of the regions (gray area) that quaternions became invalid. I graphed the pitch in two ways: integrating the gyroscope how we currently do it (blue), and using the IMU's quaternions to get orientation instead (red).
![pic1](https://github.com/user-attachments/assets/bf676490-a275-48b8-910d-70dc479b69fb)

It's clear that there are two notable "jumps" in the quaternion orientation, while the gyroscope orientation stays steady, as expected. While it's hard to draw many conclusions from the larger jump because it happened during an invalid section where we did not log data, the first jump happens around a tenth of a second before quaternions are marked invalid, so this was a quaternion with an unnatural jump that the IMU returned. The packets themselves are shown below:
![image](https://github.com/user-attachments/assets/8fca7f56-9255-43fd-83cb-5b53cd5af12b)

Both genesis launches did not exhibit any jumps similar to these during motor burn or coast phase, but they existed in freefall state. Also, take note of the precision between the two measurements up until the first jump in data.
![pic2](https://github.com/user-attachments/assets/bc305d69-8d34-401b-90da-9caccfcb593c)
![pic3](https://github.com/user-attachments/assets/e0c16571-c4a7-4dd2-8027-ac5cd0922694)

Despite not having any quaternion jumps during coast phase, I believe that this was either just luck, or a result of being smaller (slower, shorter) flights. We also cannot use quaternion uncertainty to validate quaternions, because these jumps occurred during relatively normal uncertainty levels, and they never corrected themselves either, the error only accumulated.

I think it is possible that a magnetometer could completely stop these jumps, or allow for some internal correcting in the IMU readings. Although, until we can get launch data with a magnetometer connected, I think quaternions are too volatile to use and we should stick with integrating gyroscope. There is an argument to be made for checking to see if the quaternions have jumped and then using gyroscope for the rest of the flight, but it seems like more effort than what it is worth, especially because performance has increased a lot and that would be the main benefit of not integrating gyroscope.
